### ISSUE TITLE
Different behaviours when using Rule annotation with Pact

### ISSUE THREAD

**Issue by wedla** [pact-foundation/pact-jvm#1363]
Hello, friends!

I need your help to understand an issue I got by using the @Rule annotation. Let's see what happened.

I was using @Rule in a field from a class I use extending in others, as you can see below:
```
@Rule
public PactProviderRule mockProvider = new PactProviderRule("kmv_provider", "localhost",
			RedeUtilitarios.obterPortaDisponivel(), this);
```

I wrote a pact and a test for it. This pact returns a 422 and a body, but I was validating only the http code as you can see below:

```
@Pact(consumer = "kmv_checkout_consumer")
	public RequestResponsePact criarPactoFalhaVipComoConsumidor(PactDslWithProvider construtor) throws IOException, CloneNotSupportedException {
		var acumuloPontos = RequisicaoServicoAcumuloPontos.construirObjetoVipComoConsumidor();
		var corpo = Utilitarios.obterJsonStringDoObjeto(acumuloPontos);
		
		var resposta = RespostaServicoAcumuloPontos.construirObjetoVipComoConsumidor();
		var corpoResposta = RespostaServicoAcumuloPontos.construirObjetoMatchingRules(resposta);
		
		return construtor
				.given("03 - A requisicao eh respondida com falha - CPF do VIP passado como CPF do consumidor.")
				.uponReceiving("Uma requisicao com um corpo valido, porem enviando o CPF do VIP como CPF do consumidor.")
					.path(caminho)
					.method(verbo)
					.body(corpo)
					.headers(cabecalhoRequisicao)
				.willRespondWith()
					.status(HttpStatus.UNPROCESSABLE_ENTITY.value())
					.headers(cabecalhoRequisicao)
					.body(corpoResposta)
				.toPact();
}
```
In the test class, I wrote the following method:

`@Test
@PactVerification(fragment = "criarPactoFalhaVipComoConsumidor") 
public void testeCenarioFalhaVipComoConsumidor() throws JsonParseException, JsonMappingException, IOException, CloneNotSupportedException {
		var acumuloPontos = RequisicaoServicoAcumuloPontos.construirObjetoVipComoConsumidor();
		var corpo = Utilitarios.obterJsonStringDoObjeto(acumuloPontos);
		var entidadeHttp = new HttpEntity<>(corpo, cabecalho);
		
		// then
		//expectedException.expect(HttpClientErrorException.UnprocessableEntity.class);
		
		// when
		//requisicao.exchange(**mockProvider.getUrl()** + caminho, verbo, entidadeHttp, String.class);
}`

In this way, the test gets green and all works fine. But i made a change in my base class containing the mock provider, as it follows:

private PactProviderRule mockProvider = new PactProviderRule("kmv_provider", "localhost",
			RedeUtilitarios.obterPortaDisponivel(), this);
	
@Rule
public PactProviderRule getMockProvider() {
		return this.mockProvider;
}

And in the test class, I was calling it as it follows:
`@Test
@PactVerification(fragment = "criarPactoFalhaVipComoConsumidor") 
public void testeCenarioFalhaVipComoConsumidor() throws JsonParseException, JsonMappingException, IOException, CloneNotSupportedException {
		var acumuloPontos = RequisicaoServicoAcumuloPontos.construirObjetoVipComoConsumidor();
		var corpo = Utilitarios.obterJsonStringDoObjeto(acumuloPontos);
		var entidadeHttp = new HttpEntity<>(corpo, cabecalho);
		
		// then
		//expectedException.expect(HttpClientErrorException.UnprocessableEntity.class);
		
		// when
		//requisicao.exchange(**getMockProvider().getUrl()** + caminho, verbo, entidadeHttp, String.class);
}`

As you can see, in the first method I called **mockProvider.getUrl()** and in the next I called **getMockProvider().getUrl()**. When I do the next, the exception message is also validated, even I'm not writing expectedException.expectMessage("some message") and the test gets blue with an assertion fail.

So, what is happening? Is this a correct behaviour? If a @Rule is used in a field it behaves in a way and if is used in a method it behaves differently? I confess I think it is kind of weird, so I'm opening this issue so you can avaliate this case.

Thank in advance!!

**Comment by uglyog** [comment]
I don't think this is related to the Pact libs, but maybe JUnit or Spring behaviour. Pact doesn't deal with exceptions, just the HTTP requests and responses. I have also never seen what the behaviour of the @Rule annotation on a method does, because we have always just used it on an instance.

It might change the order that things are run (i.e. the Spring test support and the Pact ones). The best thing to do is enable debug level logging and provide the logs from before and after.
